File: core/src/main/java/dev/bee/kanjianki/core/study/WritingAnalysisEngine.java
package dev.bee.kanjianki.core.study;

import java.util.Collections;
import java.util.List;

public final class WritingAnalysisEngine {
    private WritingAnalysisEngine() {
    }

    public static WritingAnalysis noInk() {
        return noInk(HintLevel.BLIND, 0);
    }

    public static WritingAnalysis noInk(HintLevel hintLevel, int hintsUsed) {
        return new WritingAnalysis(WritingAnalysis.Status.NO_INK, "again", false, "Write in the square before checking.", Collections.emptyList(), null, hintLevel, hintsUsed);
    }

    public static WritingAnalysis modelUnavailable(String message) {
        return modelUnavailable(message, HintLevel.BLIND, 0);
    }

    public static WritingAnalysis modelUnavailable(String message, HintLevel hintLevel, int hintsUsed) {
        return new WritingAnalysis(WritingAnalysis.Status.MODEL_UNAVAILABLE, "again", false, message, Collections.emptyList(), null, hintLevel, hintsUsed);
    }

    public static WritingAnalysis recognitionError(String message) {
        return recognitionError(message, HintLevel.BLIND, 0);
    }

    public static WritingAnalysis recognitionError(String message, HintLevel hintLevel, int hintsUsed) {
        return new WritingAnalysis(WritingAnalysis.Status.RECOGNITION_ERROR, "again", false, "The handwriting checker could not read this attempt. Try once more.", Collections.emptyList(), null, hintLevel, hintsUsed);
    }

    public static WritingAnalysis analyze(String target, WritingSample sample, StrokeGuide guide, List<RecognitionCandidate> candidates) {
        return analyze(target, sample, guide, candidates, HintLevel.BLIND, 0);
    }

    public static WritingAnalysis analyze(
            String target,
            WritingSample sample,
            StrokeGuide guide,
            List<RecognitionCandidate> candidates,
            HintLevel hintLevel,
            int hintsUsed
    ) {
        if (sample == null || !sample.hasInk()) {
            return noInk(hintLevel, hintsUsed);
        }
        StrokeOrderEvaluator.StrokeOrderResult order = StrokeOrderEvaluator.evaluate(guide, sample);
        if (order.missingGuide) {
            RecognitionMatch match = match(target, candidates);
            if (match.recognized) {
                String message = match.topCandidate
                        ? "Recognized as the target kanji. Stroke order could not be checked because no guide is bundled yet."
                        : "Recognized as the target kanji, but stroke order could not be checked because no guide is bundled yet.";
                return new WritingAnalysis(WritingAnalysis.Status.CLOSE, match.topCandidate ? "good" : "hard", true, message, candidates, order, hintLevel, hintsUsed);
            }
            return new WritingAnalysis(WritingAnalysis.Status.NO_STROKE_DATA, "again", false, order.message + " I could not read that as the target kanji yet.", candidates, order, hintLevel, hintsUsed);
        }
        RecognitionMatch match = match(target, candidates);
        if (!match.recognized) {
            return new WritingAnalysis(WritingAnalysis.Status.WRONG, "again", false, "I could not read that as the target kanji yet.", candidates, order, hintLevel, hintsUsed);
        }
        if (!order.acceptable) {
            return new WritingAnalysis(WritingAnalysis.Status.WRONG, "again", false, order.message, candidates, order, hintLevel, hintsUsed);
        }
        if (!order.clean) {
            order = order.withDiagnosis(order.diagnosis.plus(StrokeDiagnosis.Label.RECOGNIZED_BUT_MESSY, 0));
            return new WritingAnalysis(WritingAnalysis.Status.CLOSE, "hard", true, "Readable, but the stroke path needs one more careful pass.", candidates, order, hintLevel, hintsUsed);
        }
        if (match.topCandidate) {
            return new WritingAnalysis(WritingAnalysis.Status.PASS, "easy", true, "Clean match.", candidates, order, hintLevel, hintsUsed);
        }
        return new WritingAnalysis(WritingAnalysis.Status.PASS, "good", true, "Matched the kanji. Keep tightening the stroke path.", candidates, order, hintLevel, hintsUsed);
    }

    private static RecognitionMatch match(String target, List<RecognitionCandidate> candidates) {
        if (target == null || candidates == null || candidates.isEmpty()) {
            return new RecognitionMatch(false, false);
        }
        for (int i = 0; i < candidates.size(); i++) {
            String text = normalizedCandidate(candidates.get(i).text);
            if (target.equals(text)) {
                return new RecognitionMatch(true, i == 0);
            }
        }
        return new RecognitionMatch(false, false);
    }

    private static String normalizedCandidate(String text) {
        if (text == null) {
            return "";
        }
        return text.trim()
                .replace("\uFE0E", "")
                .replace("\uFE0F", "");
    }

    private static final class RecognitionMatch {
        final boolean recognized;
        final boolean topCandidate;

        RecognitionMatch(boolean recognized, boolean topCandidate) {
            this.recognized = recognized;
            this.topCandidate = topCandidate;
        }
    }
}

File: writing-core/src/main/kotlin/dev/bee/kanjianki/core/study/WritingAnalysisEngine.kt
package dev.bee.kanjianki.core.study

import java.util.Collections

class WritingAnalysisEngine private constructor() {
    private data class RecognitionMatch(val recognized: Boolean, val topCandidate: Boolean)

    companion object {
        private const val TARGET_NOT_RECOGNIZED = "I could not read that as the target kanji yet."
        private const val RECOGNITION_ERROR_MESSAGE = "The handwriting checker could not read this attempt. Try once more."

        @JvmStatic
        fun noInk(): WritingAnalysis = noInk(HintLevel.BLIND, 0)

        @JvmStatic
        fun noInk(hintLevel: HintLevel?, hintsUsed: Int): WritingAnalysis {
            return WritingAnalysis(
                WritingAnalysis.Status.NO_INK,
                StudyRating.AGAIN.code(),
                false,
                "Write in the square before checking.",
                Collections.emptyList(),
                null,
                hintLevel,
                hintsUsed
            )
        }

        @JvmStatic
        fun modelUnavailable(message: String?): WritingAnalysis = modelUnavailable(message, HintLevel.BLIND, 0)

        @JvmStatic
        fun modelUnavailable(message: String?, hintLevel: HintLevel?, hintsUsed: Int): WritingAnalysis {
            return WritingAnalysis(
                WritingAnalysis.Status.MODEL_UNAVAILABLE,
                StudyRating.AGAIN.code(),
                false,
                message,
                Collections.emptyList(),
                null,
                hintLevel,
                hintsUsed
            )
        }

        @JvmStatic
        fun recognitionError(): WritingAnalysis = recognitionError(HintLevel.BLIND, 0)

        @JvmStatic
        fun recognitionError(hintLevel: HintLevel?, hintsUsed: Int): WritingAnalysis {
            return WritingAnalysis(
                WritingAnalysis.Status.RECOGNITION_ERROR,
                StudyRating.AGAIN.code(),
                false,
                RECOGNITION_ERROR_MESSAGE,
                Collections.emptyList(),
                null,
                hintLevel,
                hintsUsed
            )
        }

        @JvmStatic
        fun analyze(
            target: String?,
            sample: WritingSample?,
            guide: StrokeGuide?,
            candidates: List<RecognitionCandidate>?,
        ): WritingAnalysis {
            return analyze(target, sample, guide, candidates, HintLevel.BLIND, 0)
        }

        @JvmStatic
        fun analyze(
            target: String?,
            sample: WritingSample?,
            guide: StrokeGuide?,
            candidates: List<RecognitionCandidate>?,
            hintLevel: HintLevel?,
            hintsUsed: Int,
        ): WritingAnalysis {
            if (sample == null || !sample.hasInk()) {
                return noInk(hintLevel, hintsUsed)
            }
            var order = StrokeOrderEvaluator.evaluate(guide, sample)
            if (order.missingGuide) {
                return analyzeWithoutGuide(target, candidates, order, hintLevel, hintsUsed)
            }
            val match = match(target, candidates)
            if (!match.recognized) {
                order = order.withDiagnosis(confusionDiagnosis(target, candidates, order.diagnosis))
                return WritingAnalysis(
                    WritingAnalysis.Status.WRONG,
                    StudyRating.AGAIN.code(),
                    false,
                    TARGET_NOT_RECOGNIZED,
                    candidates,
                    order,
                    hintLevel,
                    hintsUsed
                )
            }
            if (!order.acceptable) {
                return WritingAnalysis(
                    WritingAnalysis.Status.WRONG,
                    StudyRating.AGAIN.code(),
                    false,
                    order.message,
                    candidates,
                    order,
                    hintLevel,
                    hintsUsed
                )
            }
            if (!order.clean) {
                order = order.withDiagnosis(order.diagnosis.plus(StrokeDiagnosis.Label.RECOGNIZED_BUT_MESSY, 0))
                return WritingAnalysis(
                    WritingAnalysis.Status.CLOSE,
                    StudyRating.HARD.code(),
                    true,
                    "Readable, but the stroke path needs one more careful pass.",
                    candidates,
                    order,
                    hintLevel,
                    hintsUsed
                )
            }
            if (match.topCandidate) {
                return WritingAnalysis(
                    WritingAnalysis.Status.PASS,
                    StudyRating.EASY.code(),
                    true,
                    "Clean match.",
                    candidates,
                    order,
                    hintLevel,
                    hintsUsed
                )
            }
            return WritingAnalysis(
                WritingAnalysis.Status.PASS,
                StudyRating.GOOD.code(),
                true,
                "Matched the kanji. Keep tightening the stroke path.",
                candidates,
                order,
                hintLevel,
                hintsUsed
            )
        }

        private fun analyzeWithoutGuide(
            target: String?,
            candidates: List<RecognitionCandidate>?,
            order: StrokeOrderEvaluator.StrokeOrderResult,
            hintLevel: HintLevel?,
            hintsUsed: Int,
        ): WritingAnalysis {
            val match = match(target, candidates)
            if (match.recognized) {
                val message = if (match.topCandidate) {
                    "Recognized as the target kanji. Stroke order could not be checked because no guide is bundled yet."
                } else {
                    "Recognized as the target kanji, but stroke order could not be checked because no guide is bundled yet."
                }
                return WritingAnalysis(
                    WritingAnalysis.Status.CLOSE,
                    if (match.topCandidate) StudyRating.GOOD.code() else StudyRating.HARD.code(),
                    true,
                    message,
                    candidates,
                    order,
                    hintLevel,
                    hintsUsed
                )
            }
            return WritingAnalysis(
                WritingAnalysis.Status.NO_STROKE_DATA,
                StudyRating.AGAIN.code(),
                false,
                "${order.message} $TARGET_NOT_RECOGNIZED",
                candidates,
                order,
                hintLevel,
                hintsUsed
            )
        }

        private fun match(target: String?, candidates: List<RecognitionCandidate>?): RecognitionMatch {
            if (target == null || candidates.isNullOrEmpty()) {
                return RecognitionMatch(false, false)
            }
            for (i in candidates.indices) {
                val text = normalizedCandidate(candidates[i].text)
                if (target == text) {
                    return RecognitionMatch(true, i == 0)
                }
            }
            return RecognitionMatch(false, false)
        }

        private fun confusionDiagnosis(
            target: String?,
            candidates: List<RecognitionCandidate>?,
            diagnosis: StrokeDiagnosis,
        ): StrokeDiagnosis {
            val top = candidates?.firstOrNull()?.text?.let(::normalizedCandidate) ?: ""
            if (target.isNullOrEmpty() || top.isEmpty() || top == target || top.length != 1) {
                return diagnosis
            }
            return diagnosis.plus(StrokeDiagnosis.Label.CONFUSED_WITH_SIMILAR_KANJI, 0)
        }

        private fun normalizedCandidate(text: String): String {
            return text.trim()
                .replace("\uFE0E", "")
                .replace("\uFE0F", "")
        }
    }
}
